### PR TITLE
Sidebar: Better search, keyboard shortcuts, and "recently viewed"

### LINKED ISSUE #10757: Better search
**Is your feature request related to a problem? Please describe.**
Hi!

We have hundreds of components in our company's storybook, but we find that there are a lot of struggles when developers/designers are trying to find components that are relevant to them.
A common complaint during our quarterly dev survey ends up being "storybook is great, but really hard to search"

We use the custom story sort as we have a "UI" directory that we actually want to float to the top of our tree structure, as they represent our main shared component library (over say, product specific components)

However, we find that the current fuzzy search ends up being far noisier than we actually want, which means that some search queries end up being totally ineffective.

Some examples:
`dropdown` should ideally pop up our UI dropdown component, not a product specific one
image

`button` doesn't actually find `button`, and the accordion component doesn't seem to have anything that actually relates
image

Example, where we want to search for an "icon" but our actual icon component is at the bottom of the list
image

**Describe the solution you'd like**
I would like to opt-out of using the fuse fuzzy search library and implement our own search, picking and choosing the terms that we'd want to search on for each story as opposed to the defaults here https://github.com/storybookjs/storybook/blob/master/lib/ui/src/components/sidebar/treeview/utils.js#L154

Ideally, this would come in the form of a new storybook option to define my own filter function, with the note that we would have to handle our own memoization/performance.
This should more-or-less match the same function signature already provided, given a `dataset` and `filter`

Then, the custom storysort should also be provided a new parameter for `filter` to provide more information than just `a, b` in case we want to rank different search results based on the filtered value.

**Describe alternatives you've considered**
We've considered:
- rolling our own "storybook" in house (but following the storybook api for compatibility)
- renaming our stories to have more keywords/match the search better (we can't do this because we use storyshots extensively, and it would cause snapshot updates throughout our entire codebase, as well as make our storynames really noisy)

**Are you able to assist bring the feature to reality?**
yes, I can! would be happy to discuss options :)

**Additional context**

### PR DESCRIPTION
**Pull request by ghengeveld**
Issue: #10757

## What I did

This completely overhauls the sidebar. It replaces the Tree to be more performant and to add keyboard navigation, as well as expand/collapse all functionality. The Search has been replaced to show results in a list instead of filtering the tree, improving usability and performance. Furthermore I've added a "recently opened" list so allow quickly switching between recently viewed components. Additionally, this fixes a bug in composition where stories with the same path/id between refs would both get highlighted when one was selected.

Subtleties:
- Scroll into view for highlighted and selected nodes
- Allow collapsing a node even if a story is selected inside
- Double expand/collapse via keyboard does expand all / collapse all

Video: https://youtu.be/sbiYhmL2j2I

## How to test

I've added stories for every major component and state in the sidebar. Besides that, it reuses existing logic that's already covered by unit tests.

This is going to require significant QA. Things to consider:
- Navigating the tree with the mouse
- Navigating with keyboard, both with tab and arrow keys
- Expanding, collapsing nodes + collapse/expand all (try expanding an already expanded node etc)
- Making changes to the Storybook (i.e. HMR functionality), does the tree update accordingly and scroll into view?
- What happens if the selected story becomes broken?
- Enable "highlight updates" in React DevTools. Does it rerender an excessive amount of things?

## TO DO

- [x] Link up search results / recently viewed to actually open the selected result
- [x] Properly sync up search input state with Downshift state
- [x] Show the first story when clicking to expand a component for the first time
- [x] Search by and highlight the component path as well
- [ ] Design approval by @domyen 
- [x] Cleanup old stuff
- [ ] QA

### CODE REVIEW

**Comment by shilman** [comment]
@ghengeveld this is looking amazing! loving the snappy search and the recently searched list.

```md
small bug: grabs keyboard focus

1. Open a story in `official-storybook` that has controls
2. Start typing in a text field
3. Type `/`

Old behavior: adds "/" to text field
New behavior: adds "/" to text field AND sets focus to search field
```

File: lib/ui/src/components/sidebar/Explorer.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import React from 'react';
+
+import { Explorer } from './Explorer';
+import { mockDataset } from './mockdata';
+import { RefType } from './types';
+
+export default {
+  component: Explorer,
+  title: 'UI/Sidebar/Explorer',
+  excludeStories: /.*Data$/,
```
Comment (shilman) [inline lib/ui/src/components/sidebar/Explorer.stories.tsx:10]: FWIW, i don't think we really need this anymore with args

File: lib/ui/src/components/sidebar/Explorer.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    });
+
+    return (
+      <div ref={containerRef}>
+        {dataset.entries.map(([refId, ref]) => (
+          <Ref
+            {...ref}
+            key={refId}
+            isBrowsing={isBrowsing}
+            selectedId={selected && selected.refId === ref.id ? selected.storyId : null}
```
Comment (shilman) [inline lib/ui/src/components/sidebar/Explorer.tsx:32]:
```suggestion
            selectedId={selected?.refId === ref.id ? selected.storyId : null}
```

File: lib/ui/src/components/sidebar/Explorer.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            highlightedId={highlighted && highlighted.refId === ref.id ? highlighted.storyId : null}
```
Comment (shilman) [inline lib/ui/src/components/sidebar/Explorer.tsx:33]:
```suggestion
            highlightedId={highlighted?.refId === ref.id ? highlighted.storyId : null}
```

File: lib/ui/src/components/sidebar/data.test.ts [hunk 4]
```diff
@@ -32,6 +32,7 @@ const a1: Item = {
   isComponent: false,
   isRoot: false,
   parent: 'a',
+  args: [],
```
Comment (shilman) [inline lib/ui/src/components/sidebar/data.test.ts:35]:
```suggestion
  args: {},
```

???

File: lib/ui/src/components/sidebar/data.test.ts [hunk 5]
```diff
@@ -52,6 +53,7 @@ const b1: Item = {
   isRoot: false,
   isComponent: false,
   parent: 'b',
+  args: [],
```
Comment (shilman) [inline lib/ui/src/components/sidebar/data.test.ts:56]:
```suggestion
  args: {},
```

File: lib/ui/src/components/sidebar/data.test.ts [hunk 6]
```diff
@@ -62,6 +64,7 @@ const b2: Item = {
...
```
Comment (shilman) [inline lib/ui/src/components/sidebar/data.test.ts:67]:
```suggestion
  args: {},
```

File: lib/ui/src/components/sidebar/data.test.ts [hunk 7]
```diff
@@ -103,6 +106,7 @@ describe('collapse all stories', () => {
         isRoot: false,
         isComponent: true,
         isLeaf: true,
+        args: [],
```
Comment (shilman) [inline lib/ui/src/components/sidebar/data.test.ts:109]:
```suggestion
        args: {},
```

File: lib/ui/src/components/sidebar/data.test.ts [hunk 8]
```diff
@@ -114,6 +118,7 @@ describe('collapse all stories', () => {
...
```
Comment (shilman) [inline lib/ui/src/components/sidebar/data.test.ts:121]:
```suggestion
        args: {},
```

File: lib/ui/src/components/sidebar/useLastViewed.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import { window } from 'global';
+import { useCallback, useEffect, useState } from 'react';
+
+import { Selection, StoryRef } from './types';
+
+const retrieveLastViewedStoryIds = (): StoryRef[] => {
+  try {
+    const raw = window.localStorage.getItem('lastViewedStoryIds');
```
Comment (shilman) [inline lib/ui/src/components/sidebar/useLastViewed.ts:8]: FYI, we use https://www.npmjs.com/package/store2 in other parts of the code. not sure whether it matters.

File: lib/ui/src/components/sidebar/utils.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+export const prevent = (e: SyntheticEvent) => {
+  e.preventDefault();
+  return false;
+};
+
+export const createId = (id: string, prefix: string) => `${prefix}_${id}`;
+
+export const get = memoize(1000)((id: string, dataset: Dataset) => dataset[id]);
```
Comment (shilman) [inline lib/ui/src/components/sidebar/utils.ts:22]: do these memos actually improve perf?

**Review by shilman — APPROVED** [review]
Fantastic stuff @ghengeveld 👏 👏 👏

File: lib/ui/src/components/sidebar/utils.ts (hunk 10 above)
Comment (ghengeveld) [reply lib/ui/src/components/sidebar/utils.ts:22]: Not sure. They were there before: https://github.com/storybookjs/storybook/blob/next/lib/ui/src/components/sidebar/Tree/utils.ts#L58

File: lib/ui/src/components/sidebar/useLastViewed.ts (hunk 9 above)
Comment (ghengeveld) [reply lib/ui/src/components/sidebar/useLastViewed.ts:8]: 👍

File: lib/ui/src/components/sidebar/utils.ts (hunk 10 above)
Comment (ghengeveld) [reply lib/ui/src/components/sidebar/utils.ts:22]: In the newly added places they are needed to prevent unnecessary rerenders I believe.

**Comment by domyen** [comment]
Nice work Gert! 

It seems like the CI isn't working for some reason. That's preventing me from seeing the published SB and reviewing the UI. I think `next` may be working again.

**Comment by domyen** [comment]
Well done 👏! This is impressive. 

**Requests:** 
- Make sidebar wider from 210px → 230px. The width was set at 210px in 5.0. With a year of real use, that width is too tight given the length of many story names. We never got around to changing it.
- The keyboard shortcuts in the (...) menu should be styled like the [/] in the search field. [i.e., like this](https://projects.invisionapp.com/d/main#/console/20491930/430355148/preview)

**Search UI/UX:**
- Show more/all results in search. Showing so few results makes the search _seem_ underpowered (which it's not). 
- Change "Show all x results" to "Show x more results". When I click on that button I want to know how many _more_ results will be revealed. I care less about the total number of results.
- In the search results and recently opened, the `›` makes the path look like a breadcrumb when its not. Change to `/`. This aligns with how nested file structures are displayed.
- Change search cancel button right position to `right: 8px`
- Preserve search input value, when I click outside the search field it shouldn't clear the search input. When I'm in the search modality I want search results to persist until I manually clear it with the (X) button. This allows me to jump between search results to find the component I want.
- Overflows for path should also cut off at the end of the path with an (...) image
- Add the full path in the `title` attribute so the user can see the path on hover. e.g. `Storybook Design System / Design system / Avatar`


p.s., The "clear history" button is a nice detail. 👨‍🍳 💋

File: lib/ui/src/components/sidebar/Search.tsx [hunk 11]
```diff
@@ -1,145 +1,320 @@
...
+  const [inputPlaceholder, setPlaceholder] = useState('Find components');
+  const [allComponents, showAllComponents] = useState(false);
 
-  '&:hover': {
-    background: opacify(0.1, theme.appBorderColor),
-  },
-}));
+  const selectStory = useCallback(
+    (id: string, refId: string) => {
+      if (api) api.selectStory(id, undefined, { ref: refId !== DEFAULT_REF_ID && refId });
```
Comment (tooppaaa) [inline lib/ui/src/components/sidebar/Search.tsx:154]:
```suggestion
     api?.selectStory(id, undefined, { ref: refId !== DEFAULT_REF_ID && refId });
```

File: lib/ui/src/components/sidebar/Search.tsx [hunk 12]
```diff
@@ -1,145 +1,320 @@
...
-    paddingTop: 2,
-    paddingBottom: 2,
-    paddingLeft: 20,
-    paddingRight: 20,
-  },
+  useEffect(() => {
+    const focusSearch = (event: KeyboardEvent) => {
+      if (isLoading || !inputRef.current) return;
+      if (event.shiftKey || event.metaKey || event.ctrlKey || event.altKey) return;
+      if (event.key === '/' && inputRef.current !== document.activeElement) {
```
Comment (tooppaaa) [inline lib/ui/src/components/sidebar/Search.tsx:165]: There's a disable shortcut option that is not taken into account here, should it be ? (and hide shortcut in the meantime)

**Review by tooppaaa — COMMENTED** [review]
Lots of good stuff !  Hard to review everything, I might do another round sometime soon :)

**Comment by ghengeveld** [comment]
> * Make sidebar wider from 210px → 230px. The width was set at 210px in 5.0. With a year of real use, that width is too tight given the length of many story names. We never got around to changing it.

I've implemented this as the default width now, but kept the minimum width at 210px. Let me know if you want it to be a hard minimum.

> The keyboard shortcuts in the (...) menu should be styled like the [/] in the search field. i.e., like this

Missed that. Done now.

> Show more or even all results in search. Showing so few results makes the search seem underpowered (which it's not).

Changed it to show 20 results by default. I picked this number because it means the "Show more" button is visible on most desktop displays, but only at the very bottom. This is useful to gauge the total number of results without having to scroll.

> Change "Show all x results" to "Show x more results".

Good call 👍  Fixed it now.

> In the search results and recently opened, the `›` makes the path look like a breadcrumb when its not. Change to `/`. This aligns with how nested file structures are displayed.

Even though visually I like the `›` more, I agree this is better for consistency / familiarity 👍

> Change search cancel button right position to right: 8px

I had this aligned with the `[/]` icon, but you're right it looks better when moved slightly further to the right.

> Preserve search input value, when I click outside the search field it shouldn't clear the search input. When I'm in the search modality I want search results to persist until I manually clear it with the (X) button. This allows me to jump between search results to find the component I want.

Sounds like a good idea but it has some implications (both technically and functionally). Right now "recently viewed" is implemented like they're default search results, so this behavior would also apply there (unless I hack around it). Then there's the issue of focus management and keyboard listeners. When you select a result, we open the component which may claim focus and/or register keyboard listeners. That means we can lose focus on the search field and/or the ability to close the search by pressing `Escape`. I suggest we gather more feedback before going down this road.

> Overflows for path should also cut off at the end of the path with an (...)

It does, but in this case the "Design System" part is simply too short to ever get cut off because it's no longer than 50% of the total width. You can verify by inspecting the element and making the string longer.

> Add the full path in the title attribute so the user can see the path on hover. e.g. Storybook Design System / Design system / Avatar

Good call 👍 Added.

File: lib/ui/src/components/sidebar/Search.tsx (hunk 12 above)
Comment (ghengeveld) [reply lib/ui/src/components/sidebar/Search.tsx:167]: Fixed. Good call!

**Comment by shilman** [comment]
@ghengeveld it looks like the new DOM structure is breaking our cypress e2e tests:

```
  1) Navigation
       should search navigation item:
     AssertionError: Timed out retrying: Expected to find element: `#storybook-explorer-searchfield`, but never found it.
      at Context.eval (http://localhost:8001/__cypress/tests?p=cypress/integration/navigation.spec.ts:162:12)

  2) Navigation
       should display no results after searching a non-existing navigation item:
     AssertionError: Timed out retrying: Expected to find element: `#storybook-explorer-searchfield`, but never found it.
      at Context.eval (http://localhost:8001/__cypress/tests?p=cypress/integration/navigation.spec.ts:168:12)

  3) Routing
       should navigate to story addons-a11y-basebutton--default:
     AssertionError: Timed out retrying: Expected to find element: `#addons-a11y-basebutton--label`, but never found it.
      at Context.eval (http://localhost:8001/__cypress/tests?p=cypress/integration/navigation.spec.ts:175:12)
```

**Comment by domyen** [comment]
### **How many search results**
> Changed it to show 20 results by default. I picked this number because it means the "Show more" button is visible on most desktop displays, but only at the very bottom. This is useful to gauge the total number of results without having to scroll.

On further thought, I suggest removing the show more altogether (unless there are _significant_ performance benefits). This makes it easier to skim all the results without leaving the search field.

<details>
<summary>More background</summary>

This isn't a typical search where result count is useful for things like pagination or to get a sense of information scale. The pattern here is find/filter. Folks are filtering by changing the query – similar to VSCode `cmd+t` and skimming the results. They're more likely to keep the search field focused and guess different queries than click out to paginate the results. That leads me to suggest showing all the results at once to let users skim the full list instead of making them click more.

iirc, Yann mentioned this behavior while we were researching. Something to the effect of "I only care about the first few results and will keep typing until I see what I want".
</details>

### **Preserve search value**
> Right now "recently viewed" is implemented like they're default search results, so this behavior would also apply there (unless I hack around it). Then there's the issue of focus management and keyboard listeners. When you select a result, we open the component which may claim focus and/or register keyboard listeners. That means we can lose focus on the search field and/or the ability to close the search by pressing Escape. I suggest we gather more feedback before going down this road.

I don't think punting on this workflow makes sense. The current UX is intended for the ideal case. The user knows exactly what they want to click on. But in unideal cases, the flow becomes cumbersome. 

Check out this real-life scenario from Chromatic's Storybook:
1. You're working on a `PullRequests` feature
2. Search for `PullRequests` to get a list of components below (there are at least 12 results for this term)
image
3. Click on one to see it 
4. Realize that it's not the component you want 🤦‍♀️
5. Now what?

**Option: Leave behavior as is**
Folks will have to type out `PullRequests` again, then click on another result. They'll type `PullRequests` again and again to see the results because clicking on one will clear the search field. 

**Option: Change behavior to preserve search results**
Folks will click again on the search field whose value remains `PullRequests` to see the results. They click the next result to view the component. If that's not the component they're looking for, they can click on the search field again to view the results.

**Pull request merged**